### DISCUSSION TITLE
Does SQLA v2 still support DeclarativeMeta and a DeclarativeBase metaclass?

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#10972] [Usage Questions] [answered]
Hello,

With SQLA v1.4 we were able to pass a metaclass to the declarative base using the [`metaclass`](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.declarative_base.params.metaclass) argument of `declarative_base()`, and using a subclass of `DeclarativeMeta` we’d expand functionality of our mappers in controlled ways.

Is that still supported with SQLA v2 and its [`DeclarativeBase`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBase) class which, if I understand correctly, sports its own metaclass now? Should I use [`DeclarativeBaseNoMeta`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBaseNoMeta), but what’s expected from a custom metaclass? Should I continue to use [`declarative_base()`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.declarative_base) although it’s been superseded by `DeclarativeBase`?

Or should I perhaps reconsider how we implemented certain features?

Thank you!
Jens

**Comment by zzzeek** [comment, marked as answer]
everything from 1.4 is still supported, however declarative_base() is superseded by DeclarativeBase as the former cannot be supported by typing tools.

While you can also use custom metaclasses with DeclarativeBase (try using DeclarativeBaseNoMeta), there's really no reason to use custom metaclasses in modern Python, you can do everything you need using the `__init_subclass__()` method.

**Reply by jenstroeger** [reply]
[`__init_subclass__()`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) works like a charm in our case, thank you!

**Comment by gabaaa1** [comment]
Please find below an example of `__init_subclass__` that allows to rename columns and table names:

```python
from sqlalchemy.orm import Mapped, mapped_column, MappedColumn, DeclarativeBase
from sqlalchemy import String, Column, MetaData

convention = {
    "ix": f"IX_%(column_0_label)s",
    "uq": f"UQ_%(table_name)s_%(column_0_name)s",
    "ck": f"CK_%(table_name)s_%(constraint_name)s",
    "fk": f"FK_%(table_name)s_%(referred_table_name)s_%(column_0_name)s",
    "pk": f"PK_%(table_name)s",
}

table_prefix="CAT_"

def to_pascal_case(s):
    return ''.join([w.title() for w in s.split('_')])

class Base(DeclarativeBase):
    metadata = MetaData(
        schema="dbo", 
        naming_convention=convention,
    )

    def __init_subclass__(cls, **kwargs):
        # rename table name if not exists
        if hasattr(cls, "__tablename__") is False:
            tmp_name=cls.__name__
            if len(tmp_name) > 2 and tmp_name[:2] == "Db":
                tmp_name=tmp_name[2:]              
            setattr(cls, "__tablename__", f"{table_prefix}{tmp_name}s")

        # rename columns from snake_case to PascalCase
        for key, value in cls.__dict__.items():
            if isinstance(value, MappedColumn):
                if value.name is None:
                    value.column.name=to_pascal_case(key)
            elif isinstance(value, Column):
                if value.name is None:
                    value.name=to_pascal_case(key)
        super().__init_subclass__(**kwargs)

class DbRole(Base):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(400), nullable=False, unique=True)
    full_name: Mapped[str] = mapped_column(String(400), nullable=False)
```

**Reply by CaselIT** [reply]
if you don't want to have a metadata at all there is also DeclarativeBaseNoMeta https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBaseNoMeta

**Reply by gabaaa1** [reply]
Thank you for the reply, would you please show me how you would implement that on the snippet I provided. I don't understand the difference between DeclarativeBase and DeclarativeBaseNoMeta. For instance can I still declare the schema and indexes naming convention with it?

**Reply by CaselIT** [reply]
just replace ` Base(DeclarativeBase)` with ` Base(DeclarativeBaseNoMeta)`.

The metadata job is only to support the ability to add mapped attributes to a class after creation. other than that the behavior is the same

**Reply by gabaaa1** [reply]
Thank you for the explanation.
